cocotb/types/range.py

```diff
@@ -0,0 +1,173 @@
+    They have a length of 0, but the :attr:`left`, :attr:`right`, and :attr:`direction` values remain as given.
+
+    .. code-block:: python3
+
+        >>> r = Range(1, 'to', 0)  # no way to count from 1 'to' 0
+        >>> r.left, r.direction, r.right
+        (1, 'to', 0)
+        >>> r.length
+        0
+
+    .. note::
+        This is only possible when specifying the direction.
+
+    Ranges also support all the features of :class:`range` including, but not limited to:
+
+    - ``value in range`` to see if a value is in the range,
+    - ``range.index(value)`` to see what position in the range the value is,
+    - ``len(range)`` which is equivalent to :attr:`length`.
+
+    The typical use case of this type is in conjunction with :class:`~cocotb.types.Array`.
+
+    Args:
+        left: leftmost bound of range
+        direction: ``'to'`` if values are ascending, ``'downto'`` if descending
+        right: rightmost bound of range (inclusive)
+    """
+
+    __slots__ = ("_range",)
+
+    @overload
+    def __init__(self, left: int, right: int):
+        pass  # pragma: no cover
+
+    @overload
+    def __init__(self, *, left: int, right: int):
+        pass  # pragma: no cover
+
+    @overload
+    def __init__(self, left: int, direction: str, right: int):
+        pass  # pragma: no cover
+
+    def __init__(self, left, direction=None, right=None):
+        if direction is not None and right is None:
+            right, direction = direction, None
+        if direction is None:
+            # direction is 'to' if left == right
+            step = 1 if left <= right else -1
+        else:
+            direction = direction.lower()
+            if direction == "to":
+                step = 1
+            elif direction == "downto":
+                step = -1
+            else:
+                raise ValueError("direction must be 'to' or 'downto'")
+        self._range = range(left, right + step, step)
+
+    @classmethod
+    def from_range(cls, rng: range) -> "Range":
+        """Convert :class:`range` to :class:`Range`."""
+        if rng.step not in (1, -1):
+            raise ValueError("step must be '1' or '-1'")
+        obj = cls.__new__(cls)
+        obj._range = rng
+        return obj
+
+    def to_range(self) -> range:
+        """Convert :class:`Range` to :class:`range`."""
+        return self._range
+
+    @property
+    def left(self) -> int:
+        """Leftmost value in a range."""
+        return self._range.start
+
+    @property
+    def direction(self) -> str:
+        """``'to'`` if values are meant to be ascending, ``'downto'`` otherwise."""
+        return "to" if self._range.step == 1 else "downto"
+
+    @property
+    def right(self) -> int:
+        """Rightmost value in a range."""
+        return self._range.stop - self._range.step
+
+    @property
+    def length(self) -> int:
+        """Length of range."""
+        return len(self._range)
+
+    def __len__(self) -> int:
+        return self.length
```

Comment:
```suggestion
    def __len__(self) -> int:
        """Length of range."""
        return len(self._range)
```
Let's not provide two ways to spell the same thing. Python has a canonical way to ask for the length of something, let's use it.